Add a functional test for time-varying transition matrices

xjing76 opened the pull request:
Model performance test around time varying transition matrices models. 

Transition matrix with regression on seasonality 

1. Zero state 
2. Constant positive s

brandonwillard commented on tests/test_performance.py:
```diff
@@ -0,0 +1,239 @@
+        posterior_trace.posterior, var_names=["Y_t"], model=model
+    )
+
+    assert (adds_pois_ppc["Y_t"].mean(0) - y_t).std() < y_t.std() / 10
```
We should use a more "all-purpose" check here, because the relationship between a measure like this and the std. dev. of the observations will be affected by too many factors that change between tests (e.g. the changes between models, parameters, etc.)

For instance, we could check whether or not the true/observed values lie within a density/credible interval (e.g. a highest density interval from `az.hdi`) of the posterior/posterior predictive samples.

The interval approach has its shortcomings (e.g. when faced with sample results containing well separated mixtures), but it using it in/for these cases should be fine.

brandonwillard commented on tests/test_performance.py:
```diff
@@ -0,0 +1,239 @@
+    return Y_rv
+
+
+def assert_performance(model, y_t):
```
If we can use a more informative name for this helper function, that would help a lot.  As it's currently defined, this function asserts a correspondence between observations and a model's posterior predictive samples, so something along those lines would be better (e.g. `assert_posterior_predictives`).

brandonwillard commented on tests/test_performance.py:
```diff
@@ -0,0 +1,260 @@
+
+    beta = np.concatenate([[0], hour, weekday, hour_week])
+
+    MU = np.exp(1.0 + np.dot(X_tt, beta)).squeeze()
```
Why's this variable name capitalized?

brandonwillard reviewed (changes requested):
This wasn't linked to the corresponding issue, so I did that.  

Anyway, this PR adds tests with regression states (i.e. regressions outside of the time-varying dynamics), but that's not part of the original issue.  Instead, we wanted a single functional test for time-varying transition matrices; one that's actually simpler than the notebook example.

In general, we want a test that's easier to understand and contains fewer moving parts, because more moving parts makes it harder to assess only the time-varying aspects of the model.

Instead of Poisson emissions try an observation distribution with simpler emissions: e.g. `SwitchingProcess("Y_t", [pm.Constant.dist(0), pm.Constant.dist(1)], V_rv, observed=observed)`.  That example will basically result in observations that are simply the state sequences.  The resulting likelihood is about as straightforward as it gets in that scenario.

Also, we need to rethink the way we're assessing the results.  For instance, we want to confirm that transitions are actually occurring according to a properly estimated dynamic (e.g. more transitions to the second state on Tuesdays).  We can't really do that with "in-sample" posterior predictive checks, because the state estimation will necessarily be exact for such strict models with zero/non-zero emissions.  We either need to check the estimated parameters and/or use out-of-sample posterior predictive values to determine whether or not the state transition frequencies were acceptably estimated.

xjing76 commented:
I removed all the additional tests and limited to only two constant states (one zero, one positive) with no regressions within each state. Right after we are able to generate the simulation data, we split the data into training and testing for out of sample metrics. 

I added in manual setting the out of sample shape on `DiscreteMarkovChain` (`V_t`), which seems like not working. 

```
ValueError: operands could not be broadcast together with remapped shapes [original->remapped]: (4738,2,2) and requested shape (2031,2,2)
``` 

Looking into the debugger, [the gamma here](https://github.com/AmpersandTV/pymc3-hmm/blob/main/pymc3_hmm/distributions.py#L417)  attribute of  `DiscreteMarkovChain` is not getting updated. Even we [updated the shared variables](https://github.com/xjing76/pymc3-hmm/blob/tm_vary_unit_test/tests/test_performance.py#L80).

brandonwillard commented on tests/test_performance.py:
```diff
@@ -0,0 +1,149 @@
+    with theano.config.change_flags(compute_test_value="off"):
+        adds_pois_ppc = pm.sample_posterior_predictive(
+            posterior_trace.posterior.drop_vars(["Gamma", 'V_t']),
+            var_names=["V_t", "Y_t"],
```
You might need to add `Gamma` to `var_names`.

xjing76 commented:
It seems to be working pretty well on estimating the transition matrix regressions part with the current set up. 
However, the rate of actually predicting the correct state in the posterior setting is quite small. 

I think often the transition matrices being symmetric along the diagonal as in the first example. 

```
ipdb> adds_pois_ppc['V_t'][:, 30].mean()
0.46
ipdb> adds_pois_ppc['Gamma'][:, 30][0]
array([[0.9008662 , 0.0991338 ],
       [0.12714147, 0.87285853]])

ipdb> adds_pois_ppc['Gamma'][:, 150][0]
array([[0.01935663, 0.98064337],
       [0.06198774, 0.93801226]])
ipdb> adds_pois_ppc['V_t'][:, 150].mean()
0.96
```

brandonwillard commented on tests/test_performance.py:
```diff
@@ -0,0 +1,140 @@
+            model=model
+        )
+
+    assert (adds_pois_ppc['V_t'] - (test_y == 1) * 1).mean() > 0.75
```
We have the true `V_t` samples, so we should compare against that.

xjing76 replied:
Yeah, Sorry, that was not right but, `test_y` is the true impression

brandonwillard commented on tests/test_performance.py:
```diff
@@ -0,0 +1,140 @@
+
+    xi_shape = X_df.shape[1]
+
+    xi_0_true = np.array([2.0, -5.0, -3.0, 0.0, 0.0, -5.0, -5.0]).reshape(xi_shape, 1)
+    xi_1_true = np.array([-2.0, -1.0, 3.0, 0.0, 0.0, 5.0, 5.0]).reshape(xi_shape, 1)
```
We may need to use the `xi` parameters from the time-varying example notebook&mdash;or ones as similarly extreme as those; otherwise, it may be fundamentally difficult to verify good out-of-sample state predictions.

xjing76 replied:
Yes, Those that the exact parameters in the notebooks.

xjing76 commented:
I Also updated the gamma_0 to the very last record coming from the training set. But that does not really change the fact the the prediction for the states are a bit off.

xjing76 commented:
I switched to stateless transition matrix for this model test. The chances of predicting the state is being much higher.

brandonwillard commented on tests/test_performance.py:
```diff
@@ -0,0 +1,145 @@
+    return df
+
+
+def create_poisson_zero_hmm(X, mu, xis, observed):
```
This function name no longer matches the model it creates.

brandonwillard commented on tests/test_performance.py:
```diff
@@ -0,0 +1,145 @@
+    last_state = (train_y[-1] > 0) * 1
+    starting_state = [1 - last_state, last_state]
+
+    return assert_posterior_predictives(
```
As always, unless this function is called at least more than once, there's no need to make it a separate function.  

If this function is a copy of an existing function, then we should import that function.  We can also modify existing functions so that they work for more than one situation.

brandonwillard commented on tests/test_performance.py:
```diff
@@ -0,0 +1,145 @@
+            trace, var_names=["V_t", "Y_t", "Gamma"], model=model
+        )
+
+    assert np.abs(adds_pois_ppc["V_t"] - (test_y > 0) * 1).mean() < 0.3
```
Let's use the true state values instead of `(test_y > 0)`.

brandonwillard reviewed (changes requested):
Looks good; just a few minor updates needed.
